modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -1916,13 +1918,114 @@ def sort_by(
+        ):
+            # If this df is empty, we don't want to try and shuffle or sort.
+            if len(self.axes[0]) == 0 or len(self.axes[1]) == 0:
+                return self.copy()
+            # If this df only has one row partition, we don't want to do a shuffle and sort - we can
+            # just do a full-axis sort.
+            if len(self._partitions) == 1:
+                return self.apply_full_axis(
+                    1,
+                    sort_function,
+                )
+            if self.dtypes[columns[0]] == object:
+                # This means we are not sorting numbers, so we need our quantiles to not try
+                # arithmetic on the values.
+                method = "inverted_cdf"
+            else:
+                method = "linear"
+            from .utils import build_sort_functions
+
+            shuffling_functions = build_sort_functions(
+                self, columns, method, ascending, **kwargs
+            )
+            new_partitions = self._partition_mgr_cls.shuffle_partitions(
+                self._partitions,
+                shuffling_functions["sample_function"],
+                shuffling_functions["pivot_function"],
+                shuffling_functions["split_function"],
+                sort_function,
+            )
+            new_axes = self.axes
+            if kwargs.get("ignore_index", False):
+                new_axes[axis.value] = RangeIndex(len(new_axes[axis.value]))
+            else:
+                new_axes[axis.value] = self._compute_axis_labels_and_lengths(
+                    axis.value, new_partitions
+                )[0]
+            if isinstance(self.axes[axis.value], pandas.MultiIndex):
+                new_axes[axis.value] = pandas.MultiIndex.from_tuples(
+                    new_axes[axis.value].values
+                )
+            new_axes[axis.value] = new_axes[axis.value].set_names(
+                self.axes[axis.value].names
+            )
+            new_axes[axis.value ^ 1] = self.axes[axis.value ^ 1]
+            new_lengths = [None, None]
+            (
+                new_partitions,
+                new_lengths[axis.value],
+            ) = self._partition_mgr_cls.rebalance_partitions(new_partitions)
+            new_modin_frame = self.__constructor__(
+                new_partitions, *new_axes, *new_lengths, self.dtypes
+            )
+            if kwargs.get("ignore_index", False):
+                new_modin_frame._propagate_index_objs(axis=0)
+            return new_modin_frame
+        # The elif selects the case where we are only sorting by index. All partitions maintain
+        # a copy of the index for their data, so if we are just sorting by index, we can make
+        # column partitions, and sort each by their index in parallel, which should be faster than
+        # shuffling our data and sorting.
+        elif axis == Axis.ROW_WISE and len(columns) == 1 and columns[0] == "__index__":
```

Comment:
could we cover this case at the API layer? `sort_index` looks very similar: https://github.com/modin-project/modin/blob/9013f54283eb6776920ee3bf527e208a516d086d/modin/core/storage_formats/pandas/query_compiler.py#L2037

Reply:
Makes sense - I've deleted the index sort at the algebra layer!